src/client/websocket/WebSocketManager.js

```diff
@@ -37,52 +82,181 @@ class WebSocketManager extends EventEmitter {
+   * @returns {void}
    */
-  destroy() {
-    if (!this.connection) {
-      this.debug('Attempted to destroy WebSocket but no connection exists!');
+  spawn(query) {
+    if (query !== undefined) {
+      if (Array.isArray(query)) {
+        for (const item of query) {
+          if (!this.spawnQueue.includes(item)) this.spawnQueue.push(item);
+        }
+      } else if (!this.spawnQueue.includes(query)) {
+        this.spawnQueue.push(query);
+      }
+    }
+
+    if (this.spawning || !this.spawnQueue.length) return;
+
+    this.spawning = true;
+    let item = this.spawnQueue.shift();
+
+    if (typeof item === 'string' && !isNaN(item)) item = Number(item);
+    if (typeof item === 'number') {
+      const shard = new WebSocketShard(this, item, this.shards[item]);
+      this.shards[item] = shard;
+      shard.once(Events.READY, () => {
+        this.spawning = false;
+        this.client.setTimeout(() => this._handleSessionLimit(shard), 5000);
+      });
+      shard.once(Events.INVALIDATED, () => {
+        this.spawning = false;
+      });
+    } else if (item instanceof WebSocketShard) {
+      item.reconnect();
+    }
+  }
+
+  /**
+   * Creates a connection to a gateway.
+   * @param {string} [gateway=this.gateway] The gateway to connect to
+   * @returns {void}
+   */
+  connect(gateway = this.gateway) {
+    this.gateway = gateway;
+
+    if (typeof this.client.options.shards === 'number') {
+      this.debug('Spawning 1 shard');
+      this.spawn(this.client.options.shards);
+    } else if (Array.isArray(this.client.options.shards)) {
+      this.debug(`Spawning ${this.client.options.shards.length} shards`);
+      for (let i = 0; i < this.client.options.shards.length; i++) {
+        this.spawn(this.client.options.shards[i]);
+      }
+    } else {
+      this.debug(`Spawning ${this.client.options.shardCount} shards`);
+      for (let i = 0; i < this.client.options.shardCount; i++) {
+        this.spawn(i);
+      }
+    }
+  }
+
+  /**
+   * Processes a packet and queues it if this WebSocketManager is not ready.
+   * @param {Object} packet The packet to be handled
+   * @param {WebSocketShard} shard The shard that will handle this packet
+   * @returns {boolean}
+   */
+  handlePacket(packet, shard) {
+    if (packet && this.status !== Status.READY) {
+      if (!BeforeReadyWhitelist.includes(packet.t)) {
+        this.packetQueue.push({ packet, shardId: shard.id });
+        return false;
+      }
+    }
+
+    if (this.packetQueue.length) {
+      const item = this.packetQueue.shift();
+      this.client.setImmediate(() => {
+        this.handlePacket(item.packet, this.shards[item.shardId]);
+      });
+    }
+
+    if (packet && PacketHandlers[packet.t]) {
+      PacketHandlers[packet.t](this.client, packet, shard);
+    }
+
+    return false;
+  }
+
+  /**
+   * Checks whether the client is ready to be marked as ready.
+   * @returns {void}
```

Comment:
There is a `return true;` and a `return false;` in there, that's not `void`, not sure which of them is desired.